packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -188,25 +189,29 @@ export class ConnectionStateHandler {
         // Report telemetry after we set client id, but before transitioning to Connected state below!
         this.handler.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
 
-        const quorumClients = this.handler.quorumClients();
         // Check if we already processed our own join op through delta storage!
         // we are fetching ops from storage in parallel to connecting to ordering service
         // Given async processes, it's possible that we have already processed our own join message before
         // connection was fully established.
         // Note that we might be still initializing quorum - connection is established proactively on load!
-        if ((quorumClients !== undefined && quorumClients.getMember(details.clientId) !== undefined)
-            || connectionMode === "read"
-        ) {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        const alreadyProcessedOwnJoinOp = this.handler.quorumClients()?.getMember(details.clientId) !== undefined;
+        const writeConnection = connectionMode === "write";
+        if (writeConnection && !alreadyProcessedOwnJoinOp) {
             this.startJoinOpTimer();
+        } else {
+            // There should be no timer for 'read' connections,
+            // and if we processed our Join op we would have processed our previous Leave op prior to that
+            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "Unexpected timer state" */);
+
+            // For write connections, we are caught up already (since we've seen our Join op) and this will be immediate
+            // For read connections, we will wait, so "Connected" event has identical semantics to write connection
```

Comment:
I don't think "identical semantics to write connection" is entirely correct. It should explicitly mention what happens in read vs write. In both read and write, we wait to catch up to the sequence number we know about at this point in time. In write, we also wait for our own join op.
You should add this block to the big comment block before the if condition. And then maybe have comments in the if / else block specific to what happens in that scenario. Right now, its a little confusing.

Reply:
I made a bunch of changes to the comments, not sure I exactly addressed this but take a look and see if it needs further clarification?